src/System.Text.Primitives/System/Text/Encoders/UnicodeScalar.cs

```diff
@@ -0,0 +1,246 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Globalization;
+
+namespace System.Buffers.Text
+{
+    /// <summary>
+    /// Represents a 24-bit Unicode scalar value.
+    /// A scalar value is any value in the range [U+0000..U+D7FF] or [U+E000..U+10FFFF].
+    /// </summary>
+    public struct UnicodeScalar : IComparable<UnicodeScalar>, IEquatable<UnicodeScalar>
```

Comment:
Q: is `UnicodeScalar` the correct name, or should this be `CodePoint`? The two concepts are rather similar, and I'm asking because the terms are defined as:

> Code Point. (1) Any value in the Unicode codespace; that is, the range of integers from 0 to 10FFFF16. (See definition D10 in Section 3.4, Characters and Encoding.) Not all code points are assigned to encoded characters. See code point type. (2) A value, or position, for a character, in any coded character set.

> Unicode Scalar Value. Any Unicode code point except high-surrogate and low-surrogate code points. In other words, the ranges of integers 0 to D7FF16 and E00016 to 10FFFF16 inclusive. (See definition D76 in Section 3.9, Unicode Encoding Forms.)